src/clients/HubPoolClient.ts

```diff
@@ -815,140 +815,158 @@ export class HubPoolClient extends BaseAbstractClient {
+          throw new Error(
+            `Invalid runningBalances length: ${runningBalances.length}. Expected ${nTokens} or ${nTokens * 2} for chain ${
+              this.chainId
+            } transaction ${event.transactionHash}`
+          );
+        }
+        executedRootBundle.runningBalances = runningBalances.slice(0, nTokens);
+        executedRootBundle.incentiveBalances =
+          runningBalances.length > nTokens ? runningBalances.slice(nTokens) : runningBalances.map(() => toBN(0));
+        this.executedRootBundles.push(executedRootBundle);
       }
-      executedRootBundle.runningBalances = runningBalances.slice(0, nTokens);
-      executedRootBundle.incentiveBalances =
-        runningBalances.length > nTokens ? runningBalances.slice(nTokens) : runningBalances.map(() => toBN(0));
-      this.executedRootBundles.push(executedRootBundle);
     }
 
     // If the contract's current rootBundleProposal() value has an unclaimedPoolRebalanceLeafCount > 0, then
     // it means that either the root bundle proposal is in the challenge period and can be disputed, or it has
     // passed the challenge period and pool rebalance leaves can be executed. Once all leaves are executed, the
     // unclaimed count will drop to 0 and at that point there is nothing more that we can do with this root bundle
     // besides proposing another one.
-    if (pendingRootBundleProposal.unclaimedPoolRebalanceLeafCount > 0) {
-      const mostRecentProposedRootBundle = this.proposedRootBundles[this.proposedRootBundles.length - 1];
-      this.pendingRootBundle = {
-        poolRebalanceRoot: pendingRootBundleProposal.poolRebalanceRoot,
-        relayerRefundRoot: pendingRootBundleProposal.relayerRefundRoot,
-        slowRelayRoot: pendingRootBundleProposal.slowRelayRoot,
-        proposer: pendingRootBundleProposal.proposer,
-        unclaimedPoolRebalanceLeafCount: pendingRootBundleProposal.unclaimedPoolRebalanceLeafCount,
-        challengePeriodEndTimestamp: pendingRootBundleProposal.challengePeriodEndTimestamp,
-        bundleEvaluationBlockNumbers: mostRecentProposedRootBundle.bundleEvaluationBlockNumbers.map(
-          (block: BigNumber) => {
-            // Ideally, the HubPool.sol contract should limit the size of the elements within the
-            // bundleEvaluationBlockNumbers array. But because it doesn't, we wrap the cast of BN --> Number
-            // in a try/catch statement and return some value that would always be disputable.
-            // This catches the denial of service attack vector where a malicious proposer proposes with bundle block
-            // evaluation block numbers larger than what BigNumber::toNumber() can handle.
-            try {
-              return block.toNumber();
-            } catch {
-              return 0;
+    if (eventsToQuery.includes("ProposeRootBundle")) {
+      if (pendingRootBundleProposal.unclaimedPoolRebalanceLeafCount > 0) {
```

Comment:
nb. This does arguably make the HubPoolClient more brittle because a lot more of its members might not be populated. For now I'm OK with the "buyer beware" warning on tuning these events, but I do think we could improve on that.